### PR TITLE
Add datasets

### PR DESCRIPTION
**Pull request by OlivierBinette**
Add the "RLdata" synthetic datasets. These are well-known datasets with a 10% duplication rate and low level of noise. They can be deduplicated with 90%+ precision and recall using very simple linkage rules. They're often used as a sanity check for computational efficiency and disambiguation accuracy.

I also added a bipartite record linkage dataset.

### CODE REVIEW

File: mismo/_data/_datasets/union_army/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+The `CEN.csv` and `MSR.csv` datasets were derived from the [Union Army Data Set](https://www.nber.org/research/data/union-army-data-set) for use in the paper titled ["Optimal F-Score Clustering for Bipartite Record Linkage"](https://arxiv.org/pdf/2311.13923.pdf).
```
Comment (NickCrews) [inline mismo/_data/_datasets/union_army/README.md:1]: lets just drop this file, this info is already in the docstring and this is destined to get out of sync

File: mismo/datasets.py [hunk 2]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
     # parallel=False kwarg, but I'd rather just have them sorted
     # by record_id
     return backend.read_csv(path).order_by("record_id").cache()
+
+
+def load_rldata500(backend: ibis.BaseBackend | None = None) -> Table:
+    """Synthetic personal information dataset with 500 rows
```
Comment (NickCrews) [inline mismo/datasets.py:135]: Can you add in the note that you put in the PR description?

> They can be deduplicated with 90%+ precision and recall using very simple linkage rules. They're often used as a sanity check for computational efficiency and disambiguation accuracy.

I think that is useful background info for people as they are choosing datasets!

File: mismo/datasets.py [hunk 3]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
...
+        - by: int64
+          Birth year
+        - bm: int64
+          Birth month
+        - bd: int64
+          Birth day
+
+    Examples
+    --------
+    >>> load_rldata500()
```
Comment (NickCrews) [inline mismo/datasets.py:172]: nit, but I think important: Can you reorder the columns in the CSV so that record_id and label_true come first, and then the rest of the feature columns?

File: mismo/datasets.py [hunk 4]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
...
+    │         … │ …        │ …        │ …        │ …        │     … │     … │     … │          … │
+    └───────────┴──────────┴──────────┴──────────┴──────────┴───────┴───────┴───────┴────────────┘
+    """  # noqa: E501
+    path = _DATASETS_DIR / "rldata/RLdata500.csv"
+    if backend is None:
+        backend = ibis
+    return backend.read_csv(path).order_by("record_id").cache()
+
+
+def load_rldata10000(backend: ibis.BaseBackend | None = None) -> Table:
```
Comment (NickCrews) [inline mismo/datasets.py:197]: return type should be tuple[Table, Table]

Comment (NickCrews) [reply mismo/datasets.py:197]: oops, I meant this for the load_cen_msr()

File: mismo/datasets.py [hunk 5]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
...
+    │         … │ …        │ …        │ …          │ …        │     … │     … │     … │          … │
+    └───────────┴──────────┴──────────┴────────────┴──────────┴───────┴───────┴───────┴────────────┘
+    """  # noqa: E501
+    path = _DATASETS_DIR / "rldata/RLdata10000.csv"
+    if backend is None:
+        backend = ibis
+    return backend.read_csv(path).order_by("record_id").cache()
+
+
+def load_cen_msr(backend: ibis.BaseBackend | None = None) -> Table:
```
Comment (NickCrews) [inline mismo/datasets.py:260]: return type should be tuple[Table, Table]

File: mismo/datasets.py [hunk 6]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
...
+    the MSR data have a match in the CEN data. However, we can consider the linkages identified by
+    the Early Indicators of Aging project as truth. For the linking fields, we use first name, last name,
+    middle initial, and approximate birth year."
+
+    Note that the `CEN` file only contains:
+    1. The 1900 census records of soldiers who match `MSR` records in the Union Army dataset.
+    2. The 1900 census records of family members of these soldiers.
+
+    A unique identifier recorded in the `label_true` attribute can be used to identify matches. 
+    Some records have an `NA` value for this unique identifier, since they were out of the scope of the linkage performed for the Union Army Dataset.
```
Comment (NickCrews) [inline mismo/datasets.py:299]: At first I read this as some records have the literal string value "NA", and I had to go look at hte CSV to confirm. Can you rephrase as

Some records' `label_true` values are NULL, since they were out of the scope of the linkage performed for the Union Army Dataset.

?

File: mismo/datasets.py [hunk 7]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
...
+            Age of the recruit at enlistment.
+        - enlistment_date: str
+            Date of enlistment of the recruit.
+
+    Examples
+    --------
+    >>> CEN, MSR = load_cen_msr()
+    >>> CEN
+    ┏━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
+    ┃ record_id ┃ label_true ┃ last_name  ┃ first_name ┃ middle_name ┃ birth_date ┃ birth_place ┃ enlistment_age ┃ enlistment_date ┃
```
Comment (NickCrews) [inline mismo/datasets.py:355]: nit: can you reorder as first_name, middle_name, last_name?

File: mismo/datasets.py [hunk 8]
```diff
@@ -126,3 +129,272 @@ def load_patents(backend: ibis.BaseBackend | None = None) -> Table:
...
+    │         6 │  100501008 │ BYERS      │ JAMES      │ NULL        │ NULL       │ IR          │             35 │ 18610722        │
+    │         7 │  100501009 │ BEERS      │ WILLIAM    │ E           │ NULL       │ CT          │             22 │ 18610722        │
+    │         8 │  100501010 │ BENEDICT   │ THOMAS     │ E           │ NULL       │ CT          │             21 │ 18610722        │
+    │         9 │  100501011 │ CLARK      │ THEODORE   │ D           │ NULL       │ CT          │             18 │ 18610722        │
+    │         … │          … │ …          │ …          │ …           │ …          │ …           │              … │ …               │
+    └───────────┴────────────┴────────────┴────────────┴─────────────┴────────────┴─────────────┴────────────────┴─────────────────┘
+
+    >>> MSR
+    ┏━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
+    ┃ record_id ┃ last_name ┃ first_name ┃ middle_name ┃ birth_year ┃ birth_month ┃ gender ┃ birth_place ┃ label_true ┃
```
Comment (NickCrews) [inline mismo/datasets.py:374]: nit: can you reorder columns as record_id, label_true, first_name, middle_name, last_name, gender, birth_year, birth_month, birth_place?

File: mismo/datasets.py (hunk 7 above)
Comment (NickCrews) [reply mismo/datasets.py:355]: and lets make the birth dates more consistent between these two datasets? Should we use ISO format and do YYYY-MM-DD?

**Comment by NickCrews** [comment]
Thanks @OlivierBinette , I really appreciate your domain knowledge here on classic benchmark datasets to use. In general this looks really good. I had a few nits, if you don't want to fix them then I can pull and edit them myself. Most of them are about making the datasets just a bit more consistent and easier to work with in terms of column ordering and formatting. I want to make it so users have as easy of a time as possible when they use these and need as little boilerplate as possible.

File: mismo/datasets.py (hunk 6 above)
Comment (NickCrews) [reply mismo/datasets.py:299]: I am explicitly leaving NULL unquoted so people won't interpret it as a literal value.

File: mismo/datasets.py (hunk 7 above)
Comment (OlivierBinette) [reply mismo/datasets.py:360]: @NickCrews Do you want mismo to contain tools for preprocessing, data exploration, and/or checking data quality? If so, I would suggest keeping the different formats for birth dates. If not, I agree this should be standardized.

**Comment by OlivierBinette** [comment]
@NickCrews Thanks for your review! I accidentally closed this... reopened in #25.

File: mismo/datasets.py (hunk 7 above)
Comment (NickCrews) [reply mismo/datasets.py:360]: I think mismo should only have more general tools like looking at NULLs, or profiling the distribution of a column. The rather specific task of unifying date formats I don't think should be a builtin thing. I imagine the other steps of data cleaning and exploration, that mismo will help with, will still be exercised even if we unify the date formats.
